spec/codeowners/cli/main_spec.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+require 'codeowners/cli/main'
+require 'codeowners/cli/interactive_runner'
+
+# kind of integration spec to verify if all commands are properly attached
+RSpec.describe Codeowners::Cli::Main do
+  describe 'codeowners-checker check' do
+    specify do
+      runner_double = instance_double(
+        Codeowners::Cli::InteractiveRunner,
+        'validate_owners=': nil, 'default_owner=': nil, 'autocommit=': nil
+      )
+      allow(runner_double).to receive(:run_with).with(an_instance_of(Codeowners::Checker))
+      allow(Codeowners::Cli::InteractiveRunner).to receive(:new).and_return(runner_double)
+
+      expect { described_class.start ['check', '--interactive'] }.not_to(output.to_stderr)
+    end
+  end
+
+  describe 'codeowners-checker fetch' do
+    specify do
+      allow(ENV).to receive(:[]).with('GITHUB_ORGANIZATION').and_return('github')
+      allow(ENV).to receive(:[]).with('GITHUB_TOKEN').and_return('token')
+      allow(Codeowners::GithubFetcher).to receive(:get_owners).and_return(['@toptal'])
+
+      expect { described_class.start ['fetch'] }.not_to output.to_stderr
+    end
+  end
```

Comment:
This block does not seem to be an integration / functional procedure that covers the tool interactions.

<details>
  <summary>I'd propose to extract what we already have in the `checker_spec` in a shared example:</summary>

```ruby
require 'fileutils'
RSpec.shared_examples :git_project_example_setup do
  let(:folder_name) { 'project' }
  let(:from) { 'HEAD' }
  let(:to) { 'HEAD' }
  let(:git) { Git.open(folder_name, log: Logger.new(STDOUT)) }

  ENV['GITHUB_TOKEN'] = nil
  ENV['GITHUB_ORGANIZATION'] = nil

  def setup_project_folder
    on_project_folder do
      setup_code_owners
      setup_owners_list
      setup_billing_domain
      setup_shared_domain
      setup_gemfile
      setup_rubocop
    end
  end

  def on_project_folder(folder=folder_name)
    Dir.mkdir(folder) unless Dir.exist?(folder)
    Dir.chdir(folder) do
      yield
    end
  end

  def setup_code_owners
    Dir.mkdir('.github') unless Dir.exist?('.github')
    File.open('.github/CODEOWNERS', 'w+') do |file|
      file.puts <<~CONTENT
        # comment ignored
        .rubocop.yml @owner
        Gemfile @owner1
        lib/billing/* @owner2
        lib/shared/* @owner2 @owner1
      CONTENT
    end
  end

  def setup_owners_list
    File.open('.github/OWNERS', 'w+') do |file|
      file.puts <<~CONTENT
        @owner
        @owner1
        @owner2
      CONTENT
    end
  end

  def setup_billing_domain
    FileUtils.mkdir_p('lib/billing')
    File.open('lib/billing/file.rb', 'w+') do |file|
      file.puts <<~CONTENT
        # TODO: something not useful here
      CONTENT
    end
  end

  def setup_shared_domain
    FileUtils.mkdir_p('lib/shared')
    File.open('lib/shared/file.rb', 'w+') do |file|
      file.puts <<~CONTENT
        # TODO: some file that multiple owners share
      CONTENT
    end
  end

  def setup_gemfile
    File.open('Gemfile', 'w+') do |file|
      file.puts <<~CONTENT
        # frozen_string_literal: true
        source "https://rubygems.org"
        gem "granite"
      CONTENT
    end
  end

  def setup_rubocop
    File.open('.rubocop.yml', 'w+') do |file|
      file.puts <<~CONTENT
        Metrics/BlockLength:
          Path: spec/**
          Enabled: false

        Metrics/LineLength:
          Max: 120
      CONTENT
    end
  end

  def setup_git_for_project
    Git.init(folder_name)
    git.add(all: true)
    git.commit('First commit :yay:')
  end

  before do
    FileUtils.rm_r(folder_name) if Dir.exists?(folder_name)
    setup_project_folder
    setup_git_for_project
  end

  after do
    FileUtils.rm_r(folder_name)
  end
end
```

And then reuse both in `checker_spec` + `main_spec`  here.

</details>

<details>
  <summary>We can build some cases using the project definition:</summary>

```ruby
RSpec.describe Codeowners::Cli::Main do
  include_examples :git_project_example_setup

  describe '.help' do
    context 'when passing no args to the command' do
      it {expect { described_class.start([]) }.to output(/Commands:/).to_stdout }
    end

    context 'when passing the `check` command' do
      it {expect { described_class.start(["help", 'check'])}.to output(/Usage:.*check REPO/m).to_stdout }
    end
  end

  describe '.check' do
    let(:check) { described_class.start(["check", folder_name, "--from", from, "--to", to, '--interactive', 'false']) }

    context 'when no interactive mode' do
      it 'outputs successful message' do
        expect { check }
          .to exit_with_code(0)
          .and output(/File is consistent/).to_stdout
      end
    end

    context 'when add a new untracked file' do
      before do
        on_project_folder do
          filename = 'lib/new_file.rb'
          File.open(filename, 'w+') do |file|
            file.puts '# add some ruby code here'
          end
          git.add filename
          git.commit('New ruby file on lib')
        end
      end

      let(:from) { 'HEAD~1' }

      it 'fails if the file is not referenced in .github/CODEOWNERS' do
        expect {check}
          .to exit_with_code(-1)
          .and output(<<~OUTPUT).to_stdout
        File project/.github/CODEOWNERS is inconsistent:
        No owner defined
        ------------------------------
        lib/new_file.rb
        ------------------------------
        OUTPUT
      end
    end
  end
end
```
</details>

Note that the matcher for `exit_with_code`  is very important to not eat some specs as we `exit` in the main code, so it will interrupt RSpec work.
<details>
  <summary>And it's easier to test as we do on `checker_spec` we can add new files and check if it's not consistent.</summary>

```ruby
RSpec.describe Codeowners::Cli::Main do
  include_examples :git_project_example_setup

  describe '.help' do
     # ...
  end

  describe '.check' do
     # ...

    context 'when add a new untracked file' do
      before do
        on_project_folder do
          filename = 'lib/new_file.rb'
          File.open(filename, 'w+') do |file|
            file.puts '# add some ruby code here'
          end
          git.add filename
          git.commit('New ruby file on lib')
        end
      end

      let(:from) { 'HEAD~1' }

      it 'fails if the file is not referenced in .github/CODEOWNERS' do
        expect {check}
          .to exit_with_code(-1)
          .and output(<<~OUTPUT).to_stdout
        File project/.github/CODEOWNERS is inconsistent:
        No owner defined
        ------------------------------
        lib/new_file.rb
        ------------------------------
        OUTPUT
      end
    end
  end
end
```
</details>

Reply:
@mpospelov approach seems to cover that, I'm basing my changes on his branch

Reply:
@lowang sure we can definitely cover such cases with the new approach, it actually does almost the same as described :) the code is partially based on what we had in `checker_spec.rb`